datafusion/functions/src/strings.rs

```diff
@@ -124,8 +125,18 @@ pub struct StringArrayBuilder {
 
 impl StringArrayBuilder {
     pub fn with_capacity(item_capacity: usize, data_capacity: usize) -> Self {
-        let mut offsets_buffer =
-            MutableBuffer::with_capacity((item_capacity + 1) * size_of::<i32>());
+        // MutableBuffer::with_capacity will round up to multiples of `ALIGNMENT`
+        let capacity = std::cmp::max(
```

Comment:
The idea of the `with_capacity` is to allocate enough space for `item_capacity + 1` offsets, each of which is an `i32`

The code below then assumes that it has successfully allocated space for that many offsets and uses the `unsafe` API assuming that the allocation is sized for `item_capacity + 1` items

In the case where the `capacity` calculation overflows, that assumption is violated. 

Thus, I think the right fix is for the code in DataFusion to check that the value it passes to `mutable` is in fact sufficient to hold `item_capacity + 1` `i32` values

I think you can do this by using https://doc.rust-lang.org/std/primitive.i32.html#method.checked_mul instead of `*` and panic'ing if the calculation is bad

Reply:
I overthought things. Fixed.